Source/Classes/PINSpeedRecorder.m

```diff
@@ -66,69 +77,113 @@ - (instancetype)init
     return self;
 }
 
-- (void)addTaskBPS:(float)bytesPerSecond endDate:(NSDate *)endDate
+- (void)processMetrics:(NSURLSessionTaskMetrics *)metrics forTask:(NSURLSessionTask *)task host:(NSString *)host
 {
-    //if bytesPerSecond is less than or equal to zero, ignore.
-    if (bytesPerSecond <= 0) {
+    NSDate *requestStart = [NSDate distantFuture];
+    NSDate *firstByte = [NSDate distantFuture];
+    NSDate *requestEnd = [NSDate distantPast];
+    int64_t contentLength = task.countOfBytesReceived;
+    
+    for (NSURLSessionTaskTransactionMetrics *metric in metrics.transactionMetrics) {
+        if (metric.requestStartDate == nil || metric.responseStartDate == nil) {
+            //Only evaluate requests which completed their first byte.
+            return;
+        }
+        if ([requestStart compare:metric.requestStartDate] != NSOrderedAscending) {
+            requestStart = metric.requestStartDate;
+        }
+        if ([firstByte compare:metric.responseStartDate] != NSOrderedAscending) {
+            firstByte = metric.responseStartDate;
+        }
+        if ([requestEnd compare:metric.responseEndDate] != NSOrderedDescending) {
+            requestEnd = metric.responseEndDate;
```

Comment:
Nit: This could be rephrased as `requestStart = [requestStart earlierDate:metric.requestStartDate];` etc.